### DISCUSSION TITLE
Theme Switch widget

### DISCUSSION THREAD

**Discussion by madodig** [TriliumNext/Trilium#4393] [Show and tell]
Hi,

I've shared the following Theme Switch widget: https://github.com/madodig/trilium-widget-theme-switch

The widget enables easy switching between themes and (optionally) automatic switching between light/dark mode based on the time of the day.

It is possible to:
- change the appearance of the widget button:
- change the icon for the button and the dropdown items;
- choose whether the Trilium frontend reload is triggered to activate the new theme;
- choose whether to show the dropdown list of available themes or just show the button to toggle between current theme's available modes;
define a time range in which the light and dark mode should be activated.

DEMO:

https://github.com/zadam/trilium/assets/14966219/a402fbb2-97c8-494e-ab4a-121d3c22dc47

Thanks to @zadam for helping me with updating Option / Theme via private API.

**Comment by Nriver** [comment]
Wow, your work is truly brilliant. This makes theme switch so easy. I've added it to https://github.com/Nriver/awesome-trilium

**Comment by meichthys** [comment]
Very nice! Great work 👍

**Comment by dcooper94** [comment]
Love this! im a noob with Trilium, im not using the desktop version, how do i import this?

**Reply by madodig** [reply]
Hi!
You would install it just as any other widget, by importing the ZIP file.
Download the [latest release](https://github.com/madodig/trilium-widget-theme-switch/releases/) and import it to your Trilium instance (right click on a parent note where you want this note to be placed, click on "Import into note", choose the downloaded ZIP file, uncheck "Safe import" option and click on "Import". Then choose the Theme Switch note, select "Owned Attributes" (second icon from the left below the note title) and type "#widget" (without the quotes) to the end of the attributes list. Then reload the Trilium frontend (F5 or Ctrl+R).

Make sure that the self-hosted deployment is properly installed (the recommended way to deploy Trilium on servers is running it as a docker image), otherwise widgets (in general) might not work.

**Comment by dcooper94** [comment]
Hey Madodig

Thanks for this, I was able to import the switch and set it as a widget. However I am running into this error when trying to select the switch or theme
You're passing an async function to api.runOnBackend() which will likely not work as you intended. Either make the function synchronous (by removing 'async' keyword), or use api.runAsyncOnBackendWithManualTransactionHandling(

Are you able to help

**Comment by dcooper94** [comment]
@madodig  I get this error and now i cannot open my notes 
![trilium notes error](https://github.com/zadam/trilium/assets/147366361/e9f2e467-390a-4e7f-80d9-17d086a90c5e)

**Reply by madodig** [reply]
The error you get is a result of changes in recent Trilium versions which have been addressed in the latest widget release (1.2.0).
I tested it on Trilium v0.62.6, self-hosted deployment and desktop (Windows), and I don't see that error.

Did you install the latest widget release? Do you maybe still have the old release running? Which Trilium version are you running? Is there any other widget that might cause this error?

1) If you can't see anything in Trilium (due to broken note or script), read [this](https://github.com/zadam/trilium/wiki/Troubleshooting) (specifically "Broken note crashes Trilium" and "Broken script prevents application startup") and proceed with step 2
2) Go through widget's "lib" note (child of the main "Theme Switch" note) and look for occurences of api.runOnBackend and replace them with api.runAsyncOnBackendWithManualTransactionHandling. You can delete the "installer" note.
3) Reload Trilium, or if you are running it in safe mode, restart the app (desktop) or stop the service and run it normally (self-hosted).

I hope this helps...
